File: 0144-binary-tree-preorder-traversal/iterative-depth-first-search/python/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        stack = [root]
        
        while stack:
            cur = stack.pop()
            if cur:
                result.append(cur.val)
                stack.append(cur.right)    # Push right first so that left sits on top and gets popped/processed first
                stack.append(cur.left)
        return result

File: 0144-binary-tree-preorder-traversal/iterative-depth-first-search/java/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;

        Deque<TreeNode> stack = new ArrayDeque<>();
        stack.push(root);

        while (!stack.isEmpty()) {
            TreeNode cur = stack.pop();
            if (cur != null) {
                result.add(cur.val);
                if (cur.right != null)  stack.push(cur.right);    // Push right first so that left sits on top and gets popped/processed first
                if (cur.left != null)   stack.push(cur.left);
            }
        }
        return result;
    }
}
